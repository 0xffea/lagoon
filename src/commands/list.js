// @flow

import { take, put, go, chan, takeAsync } from 'js-csp';
import { contains } from 'ramda';
import gql from '../gql';
import {
  effectHandler,
  stdoutEffect,
  stderrEffect,
  gqlEffect,
  stateEffect,
  connectPut,
} from '../effects';

import typeof { Channel } from 'js-csp/es/impl/channels';
import typeof { default as Yargs } from 'yargs';
import type { BaseArgs } from './index';

const name = 'list';
const description = 'Lists specific deployment information';

export async function setup(yargs: Yargs) {
  return yargs
    .usage(`$0 ${name} [target] - ${description}`)
    .options({
      sitegroup: {
        alias: 's',
        demand: false,
        describe: 'Overrides the currently configured sitegroup (.amazeeio.yml)',
      },
    })
    .example(`$0 ${name} sites`, 'Lists all sites for the specific sitegroup configured in your .amazeeio.yml config file')
    .example(`$0 ${name} sites -s mysitegroup`, 'Lists all sites for a specific sitegroup (instead of using the config file)')
    .argv;
}


function resolveChannel(channel: Channel): Promise<number> {
  return new Promise((res, rej) => {
    takeAsync(channel, (val) => res(val));
  });
}

type Args = BaseArgs & {
  sitegroup: ?string,
  target: string,
};

export async function run(args: Args): Promise<number> {
  const mainArgs = {
    sitegroup: args.sitegroup || args.config.sitegroup,
    target: args.target,
  };

  const end = go(main, [mainArgs]);

  return resolveChannel(end);
}

type MainArgs = {
  sitegroup: string,
  target: string,
};

export function* main(args: MainArgs): Generator<*, *, *> {
  const input = chan();
  const out = chan();

  // Runs the effect-handler as long as input is open
  go(effectHandler, [input, out]);

  const log = connectPut(input, stdoutEffect);
  const logErr = connectPut(input, stderrEffect);
  const gqlQuery = connectPut(input, gqlEffect);
  const getState = connectPut(input, stateEffect);

  const {
    sitegroup,
    target,
  } = args;

  if (!contains(target, ['sites'])) {
    yield logErr(`Unknown target ${target}`);
    input.close();
    return 1;
  }

  const query = gql`query myQuery($sitegroup: String!) {
    siteGroupByName(name: $sitegroup) {
      gitUrl
      sites(first: 1000) {
        edges {
          node {
            siteName
            siteBranch
            siteEnvironment
            siteHost
            serverInfrastructure
            serverIdentifier
            serverNames
          }
        }
      }
    }
  }`;

  yield gqlQuery(query, { sitegroup });
  const result = yield take(out);

  yield log(result);

  yield getState(0);
  const { code, stack } = yield take(out);

  yield log(`State code: ${code}`);
  yield log(stack);

  input.close();
  return code;
}

export default {
  name,
  description,
  run,
};
