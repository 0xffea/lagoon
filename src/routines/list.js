// @flow

import { take, put, go, chan, takeAsync, spawn } from 'js-csp'; // eslint-disable-line
import { contains } from 'ramda';
import gql from '../gql';

import {
  stdoutEffect,
  stderrEffect,
  gqlEffect,
  stateEffect,
  connectPut,
  effectHandler,
} from '../effects';

import type { EffectHandlerArgs } from '../effects';
import typeof { Channel } from 'js-csp/es/impl/channels';
import typeof RunGQLQueryFn from '../query';

type ListSitesArgs = {
  input: Channel,
  out: Channel,
};

export function* listSites(args: ListSitesArgs): Generator<*, *, *> {
}

function* end(input: Channel, out: Channel, code: number): Object {
  yield put(input, stateEffect(code));
  const state = yield take(out);

  input.close();
  out.close();

  return state;
}

type CliRoutineArgs = EffectHandlerArgs & {
  routine: Generator<*, *, *>,
};

export function* runRoutine(args: CliRoutineArgs) {
  const { routine } = args;

  go(effectHandler, [args]);
  return go(routine, [args]);
}

type ListArgs = {
  input: Channel,
  out: Channel,
  sitegroup: string,
  target: 'sites',
};

export default function* list(args: ListArgs): Generator<*, *, *> {
  const {
    input,
    out,
    sitegroup,
    target,
    runGQLQueryFn,
  } = args;

  const log = connectPut(input, stdoutEffect);
  const logErr = connectPut(input, stderrEffect);
  const gqlQuery = connectPut(input, gqlEffect);
  const getState = connectPut(input, stateEffect);

  if (!contains(target, ['sites'])) {
    yield logErr(`Unknown target '${target}'`);
    return yield take(go(end, [input, out, 1]));
  }

  const query = gql`query myQuery($sitegroup: String!) {
    siteGroupByName(name: $sitegroup) {
      gitUrl
      sites(first: 1000) {
        edges {
          node {
            siteName
            siteBranch
            siteEnvironment
            siteHost
            serverInfrastructure
            serverIdentifier
            serverNames
          }
        }
      }
    }
  }`;

  yield gqlQuery(query, { sitegroup });
  const result = yield take(out);

  yield log(result);

  yield getState(0);
  const { code, stack } = yield take(out);

  yield log(`State code: ${code}`);
  yield log(stack);

  input.close();
  return code;
}
