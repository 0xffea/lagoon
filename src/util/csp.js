// @flow

import { takeAsync } from 'js-csp';
import { fromColl } from 'js-csp/lib/csp.operations';

import typeof { Channel } from 'js-csp/es/impl/channels';

// TODO: Make Channel generic, so we at least tell which resolve type Promise yields
export function resolveChannel(channel: Channel, keepOpen?: boolean = false): Promise<mixed|Error> {
  return new Promise((res, rej) => {
    takeAsync(channel, (val) => { 
      res(val);

      if (keepOpen) {
        channel.close();
      }
    });
  });
}

export function maybeResolveChannel(channel: Channel, keepOpen?: boolean = false): Promise<mixed> {
  return new Promise((res, rej) => {
    takeAsync(channel, (val) => { 
      if (val instanceof Error) {
        rej(val);
      }
      else {
        res(val);
      }

      if (keepOpen) {
        channel.close();
      }
    });
  });
}


export function fail(msg: string): Channel {
  return fromColl([new Error(msg)]);
}
