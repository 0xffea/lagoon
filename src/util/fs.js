// @flow

/**
 * This module exists to promisify all the fs functionality
 */

import fs from 'fs';

import type { Stats } from 'fs';

export function readFile(
  filename: string,
  encOrOpts?: string | Object,
): Promise<string | Buffer> {
  return new Promise((resolve, reject) => {
    fs.readFile(filename, encOrOpts || {}, (err, data) => {
      if (err) {
        reject(err);
      }
      else {
        resolve(data);
      }
    });
  });
}

export function writeFile(
  filename: string,
  data: Buffer | string,
  options?: Object | string,
): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.writeFile(filename, data, options, err => {
      if (err) {
        reject(err);
      }
      else {
        resolve();
      }
    });
  });
}

export function lstat(path: string): Promise<Stats> {
  return new Promise((resolve, reject) => {
    fs.lstat(path, (err, stat) => {
      if (err) {
        reject(err);
      }
      else {
        resolve(stat);
      }
    });
  });
}

export async function doesFileExist(file: string): Promise<boolean> {
  try {
    const stats = await lstat(file);

    if (stats.isFile()) {
      return true;
    }

    return false;
  }
  catch (err) {
    return false;
  }
}

export const statSync = fs.statSync;
