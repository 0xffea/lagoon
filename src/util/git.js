// @flow

import Git from 'nodegit';

import type {
  Repository,
  Signature,
  Revparse,
  Remote,
  CredAcquireCb as CredCb,
} from 'nodegit';

export type {
  Repository,
  CredAcquireCb as CredCb,
  Remote,
  Signature,
} from 'nodegit';

export function createSignature(
  time?: number = parseInt(Date.now() / 1000, 10),
  offset?: number = new Date().getTimezoneOffset()
): Signature {
  return Git.Signature.create('API', 'api@amazee.io', time, offset);
}

export function createCredentialsCb(
  username: string,
  password: string
): CredCb {
  return () => {
    return Git.Cred.userpassPlaintextNew(username, password);
  };
}

function expectDefaultState(repository: Repository): Repository {
  if (!repository.isDefaultState()) {
    throw new Error('The repository is not in the default state.');
  }

  return repository;
}

export async function getRepository(
  url: string,
  branch: string,
  destination: string,
  credCb: CredCb
): Promise<Repository> {
  try {
    // Get the repository if it already exists.
    return await Git.Repository.open(destination);
  }
  catch (e) {
    // Repository doesn't exist locally yet. Clone it.
    return await Git.Clone.clone(url, destination, {
      checkoutBranch: branch,
      fetchOpts: {
        callbacks: { certificateCheck: () => 1, credentials: credCb },
      },
    });
  }
}

export async function getRemote(
  repository: Repository,
  remote: string
): Promise<Remote> {
  return await repository.getRemote(remote);
}

export async function remotePush(
  remote: Remote,
  refs: Array<string>,
  credCb: CredCb
): Promise<void> {
  // Attempt to push any pending commits.
  await remote.push(refs, {
    callbacks: { certificateCheck: () => 1, credentials: credCb },
  });
}

export async function revparseSingle(
  repository: Repository,
  spec: string
): Promise<Revparse> {
  return await Git.Revparse.single(repository, spec);
}

export async function fetchAll(
  repository: Repository,
  credentialsCb: CredCb
): Promise<void> {
  // Fetch any changes from the remote.
  await repository.fetchAll({
    callbacks: { certificateCheck: () => 1, credentials: credentialsCb },
  });
}

// Open the repository directory or fetch it from the remote.
export async function ensureRepository(
  url: string,
  branch: string,
  destination: string,
  credCb: CredCb
): Promise<Repository> {
  const repository = await getRepository(url, branch, destination, credCb);

  // Ensure that we are on the right branch.
  await repository.checkoutBranch(branch);

  return expectDefaultState(repository);
}

export async function rebase(
  repository: Repository,
  branch: string,
  upstream: string,
  onto: string,
  sig: Signature
): Promise<void> {
  await repository.rebaseBranches(branch, upstream, onto, sig);
}

