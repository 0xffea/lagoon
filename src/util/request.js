// @flow

import https from 'https';
import defer from './defer';

export type RequestOptions = {
  hostname: string,
  path?: string,
  port?: number | string,
  method: 'POST' | 'GET',
  headers?: Object,
  rejectUnauthorized?: boolean,
  body?: string,
};

export default function request(options: RequestOptions): Promise<Object> {
  const { body } = options;

  const def = defer();

  const req = https.request(options, (res) => {
    res.setEncoding('utf8');

    let rawData = '';
    res.on('data', chunk => { rawData += chunk; });
    res.on('end', () => {
      try {
        let parsed = JSON.parse(rawData);
        def.resolve(parsed);
      }
      catch (e) {
        def.reject(e);
      }
    });
  });

  if (body) {
    req.write(body);
  }

  req.end();

  req.on('error', e => def.reject(e));

  return def.promise;
}
