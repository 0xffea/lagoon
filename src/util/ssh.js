// @flow

import sshpk from 'sshpk';

export function isKeyValid(key: string): boolean {
  try {
    // Validate the format of the ssh key. This fails with an exception
    // if the key is invalid. We are not actually interested in the
    // result of the parsing and just use this for validation.
    sshpk.parseKey(key, 'ssh');
    return true;
  } catch (e) {
    return false;
  }
}
