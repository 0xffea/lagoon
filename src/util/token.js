// @flow

import jwt from 'jsonwebtoken';

// TODO:
// Currently, this generates the same bearer token every time you
// call this with the same ssh key. Eventually, we might want to
// support multiple bearer tokens when e.g. logging in from different
// devices using the ssh key.
export function generateTokenFromKey(key: string): string {
  return jwt.sign(key, 'super-secret-string');
}
