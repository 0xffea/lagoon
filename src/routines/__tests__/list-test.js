// @flow

import { path } from 'ramda';
import { chan, takeAsync, go, put, take } from 'js-csp';
import { resolveChannel } from '../../util/csp';
import list from '../list';

import typeof { Channel } from 'js-csp/es/impl/channels';

const getEffect = (ret) => path(['value', 'value'], ret);
const getChannel = (ret) => path(['value', 'channel'], ret);

describe('list', () => {
  it('should list all sites of given sitegroup', () => {
    const input = chan();
    const out = chan();

    const iter = list({
      input,
      out,
      sitegroup: 'amazee_io',
      target: 'sites',
    });

    let ret = iter.next();

    // Asks for a graphQL result
    expect(getChannel(ret)).toEqual(input);
    expect(getEffect(ret).type).toEqual('gql');
    expect(getEffect(ret).query).toMatchSnapshot();
    expect(getEffect(ret).variables).toEqual({
      sitegroup: 'amazee_io',
    });

    // Retrieve the graphQL result
    ret = iter.next();
    expect(getChannel(ret)).toEqual(out);

    // Output the query result to stdout
    ret = iter.next();
    expect(getChannel(ret)).toEqual(input);
    expect(getEffect(ret).type).toEqual('stdout');

    // Retrieve the state and return the result code
    ret = iter.next();
  });

  it.skip('should list all sites of given sitegroup', () => {
    const input = chan();
    const out = chan();
    const runGQLQueryFn = jest.fn(() => ({
      some: 'data',
    }));

    const args = {
      input,
      out,
      sitegroup: 'amazee_io',
      target: 'sites',
      runGQLQueryFn,
    };

    const routine = go(list, [args]);

    return resolveChannel(routine).then(code => {
      expect(code).toEqual(0);
    });
  });

  it.skip('should error out on unknown target', () => {
    const input = chan();
    const out = chan();
    const runGQLQueryFn = jest.fn(() => ({
      some: 'data',
    }));

    const args = {
      input,
      out,
      sitegroup: 'amazee_io',
      target: 'unknown',
      runGQLQueryFn,
    };

    const routine = go(list, [args]);

  });
});
