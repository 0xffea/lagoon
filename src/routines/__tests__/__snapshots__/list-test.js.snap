exports[`list should list all sites of given sitegroup 1`] = `
"query myQuery($sitegroup: String!) {
    siteGroupByName(name: $sitegroup) {
      gitUrl
      sites(first: 1000) {
        edges {
          node {
            siteName
            siteBranch
            siteEnvironment
            siteHost
            serverInfrastructure
            serverIdentifier
            serverNames
          }
        }
      }
    }
  }"
`;
